File: libs/power-log-parser/src/lib/parsers/hero-revealed-parser.ts
import { CardType, GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { FullEntity, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class HeroRevealedParser implements ActionParser {
	readonly ParserName = 'HeroRevealedParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		let fullEntity: FullEntity;
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.FullEntity &&
			(fullEntity = node.Object as FullEntity).GetTag(GameTag.ZONE) === (Zone.PLAY as number) &&
			fullEntity.GetCardType() === (CardType.HERO as number)
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const fullEntity = node.Object as FullEntity;
		const cardId = fullEntity.CardId;
		const controllerId = fullEntity.GetEffectiveController();
		const health = fullEntity.GetTag(GameTag.HEALTH);
		return [
			GameEventProvider.Create(
				fullEntity.TimeStamp,
				'HERO_REVEALED',
				GameEventHelper.CreateProvider('HERO_REVEALED', cardId, controllerId, fullEntity.Id, this.StateFacade, {
					Health: health,
				}),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/HeroRevealedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class HeroRevealedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public HeroRevealedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            FullEntity fullEntity;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(FullEntity)
                && (fullEntity = node.Object as FullEntity).GetTag(GameTag.ZONE) == (int)Zone.PLAY
                && fullEntity.GetCardType() == (int)CardType.HERO;

        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var fullEntity = node.Object as FullEntity;
            var cardId = fullEntity.CardId;
            var controllerId = fullEntity.GetEffectiveController();
            var health = fullEntity.GetTag(GameTag.HEALTH);
            return new List<GameEventProvider> { GameEventProvider.Create(
                fullEntity.TimeStamp,
                "HERO_REVEALED",
                GameEvent.CreateProvider(
                    "HERO_REVEALED",
                    cardId,
                    controllerId,
                    fullEntity.Id,
                    StateFacade,
                    //null,
                    new {
                        Health = health,
                    }
                ),
                true,
                node) };
        }
    }
}
